clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -1772,18 +1872,17 @@ private void close(long timeoutMs, boolean swallowException) {
      * @throws org.apache.kafka.common.errors.AuthenticationException if authentication fails. See the exception for more details
      * @throws NoOffsetForPartitionException If no offset is stored for a given partition and no offset reset policy is
      *             defined
-     * @return true if all assigned positions have a position, false otherwise
+     * @return true iff the operation completed without timing out
      */
-    private boolean updateFetchPositions() {
-        if (subscriptions.hasAllFetchPositions())
-            return true;
+    private boolean updateFetchPositions(final long timeoutMs) {
```

Comment:
This is a bit of a meta comment. We have structured all of these internal APIs to use a timeout. It makes sense in isolation, but when you have a sequence of operations, it's annoying to repeatedly recompute the remaining time. I wonder what it would look like to structure them in terms of deadlines instead. So we just compute a deadline at the start of the call in `KafkaConsumer` and carry it through each operation.

The other thing is that we seem to have more calls to system time than we had before. For example, there are three in `internalUpdateAssignmentMetadataIfNeeded` which would hit every call to poll(). It may not seem like a big deal, but we have actually found it to be a performance, which is why we do seemingly silly optimizations like passing around the current time. Consider a case where someone configures `max.poll.records` to 1, then every record return to the user will be accompanied by a bunch of system calls. It would be nice to reduce these calls at least for the common path where we have all our offsets and we are just awaiting fetches.

Reply:
The deadline thing would be more elegant, but I'm concerned that, carelessly applied, it would spider through the clients code base; I think I'd wind up changing much more than I strictly have to. I'm not opposed, and I'll go ahead and give it a crack to see what it looks like, just wanted to point this out.

Reply:
About the extra calls to system time... Yeah, I did that deliberately because it made the code simpler, but I didn't consider that the call itself could become a bottleneck, which is completely plausible. I'll switch it back to the prior strategy, of only checking the time when it really matters.

Applying the deadline-instead-of-timeout approach would also save one call to system time at the start of every method to get the "startTime".